### DISCUSSION TITLE
cannot rotate pages of some PDF

### DISCUSSION THREAD

**Discussion by shin1007** [pymupdf/PyMuPDF#1378] [Looking for help] [answered]
**Problem description.**
In some pages of PDFs, their rotation matrix and derotation matrix are like
    Matrix(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
    Matrix(1.0, -0.0, -0.0, 1.0, 0.0, 0.0)
and we cannot rotate these pages properly.

**Solution I'd like**
adding new methods like `set_rotation_matrix` and `set_derotation_matrix`.
which set desirable matrixes and thus we can rotate pages with improper rotation matrix.

**Alternatives I've considered**
names of the methods can be different.

**Comment by JorjMcKie** [comment]
I don't understand:
Page (de-) rotation matrix is computed automatically by inspecting a page's rotation.
Changing / setting page rotation via `page.set_rotation(degree)` will also change those matrices.

**Comment by JorjMcKie** [comment]
Look at this:
```python
>>> for degree in (0, 90, 180, 270):
	page.set_rotation(degree)
	print("de-rot:", page.derotation_matrix)
	print("   rot:", page.rotation_matrix)

	
de-rot: Matrix(1.0, -0.0, -0.0, 1.0, 0.0, 0.0)
   rot: Matrix(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)
de-rot: Matrix(0.0, -1.0, 1.0, 0.0, -0.0, 842.0)
   rot: Matrix(0.0, 1.0, -1.0, 0.0, 842.0, 0.0)
de-rot: Matrix(-1.0, -0.0, -0.0, -1.0, 595.0, 842.0)
   rot: Matrix(-1.0, 0.0, 0.0, -1.0, 595.0, 842.0)
de-rot: Matrix(0.0, 1.0, -1.0, 0.0, 595.0, -0.0)
   rot: Matrix(0.0, -1.0, 1.0, 0.0, 0.0, 595.0)
>>>  
```

**Comment by shin1007** [comment]
I'll check my PDF again. thanks

**Comment by shin1007** [comment]
Thanks to your advice, I've understood it's not related to rotation matrix.
however, I'm still facing a problem with rotation.
I would be pretty happy if you kindly spare some time for solving this problem.

I'm doing a kind of 2-ups and it works for some pdf and does not work for some others.
I need to rotate the page if its paper size is a4 portrait before attaching to a3 paper left / right.
code below sometimes does not rotate as i expect.
it shows just left part of the original page(a4-landscape).

it looks working if i simply overwrite src_pdf.
but still having problem when i try to 2-up again.

are there any solutions you come up with?
TIA

```Python
a3_w, a3_h = fitz.paper_size("a3-l")
rect_a3_left = fitz.Rect(0, 0, a3_w/2, a3_h)
rect_a3_right = fitz.Rect(a3_w/2, 0, a3_w, a3_h)

a4_w, a4_h = fitz.paper_size("a4")
rect_a4_portrait = fitz.Rect(0, 0, a4_w, a4_h)
rect_a4_landscape = fitz.Rect(0, 0, a4_h, a4_w)

page = new_pdf.new_page(-1, width = a3_w, height = a3_h)

if paper_size == 'a4_landscape':
    print(paper_size_of(src_pdf[i]))
    print(src_pdf[i].rotation)
    print(src_pdf[i].rotation_matrix)
    print(src_pdf[i].derotation_matrix)
    if src_pdf[i].rotation == 0:
        src_pdf[i].set_rotation(270)
    else:
        src_pdf[i].set_rotation(src_pdf[i].rotation - 90)
    print(paper_size_of(src_pdf[i]))
    print(src_pdf[i].rotation)
    print(src_pdf[i].rotation_matrix)
    print(src_pdf[i].derotation_matrix)
    page.show_pdf_page(rect_a3_left, src_pdf, i)
```

**Comment by JorjMcKie** [comment]
The simplest way is probably this: set the source page's rotation to 0 before inserting it.
```python
src = fitz.open("source.pdf")
rect = fitz.paper_rect("a3-l")
w, h = fitz.paper_size("a3-l")
rl = fitz.paper_rect("a4")  # equal left side of A3 landscape
rr = rl + (rl.width, 0, rl.width, 0)  # the right side
doc = fitz.open()
for srcpage in src:
	srcpage.set_rotation(0)
	if not srcpage.number % 2:  # make new output page on even input number
		page = doc.new_page(width=w, height=h)
		page.show_pdf_page(rl, src, srcpage.number)
	else:
		page.show_pdf_page(rr, src, srcpage.number)
```

**Reply by shin1007** [reply]
Thank you for your quick reply.

I'm sorry but I made a mistake describing my project.
I am trying to rotate A4 LANDSCAPE page.(not portrait) 

In some PDFs i can rotate it properly but not for others.

> I'm doing a kind of 2-ups and it works for some pdf and does not work for some others.
> I need to rotate the page if its paper size is a4 *portrait* before attaching to a3 paper left / right.
> code below sometimes does not rotate as i expect.
> it shows just left part of the original page(a4-landscape).

**Comment by JorjMcKie** [comment]
If a page is landscape, then `page.rect.width > page.rect.height` **_and_** `page.rotation == 0`. If so, then do `page.set_rotation(90)` or -90 depending on how you want to see the output.

**Reply by shin1007** [reply]
I seem not making myself clear.
I did the `page.set_rotation` and it seems not working properly

1) original PDF, a4 landscape
2x ![image](https://user-images.githubusercontent.com/51842098/140904316-7a681d8f-1d96-48ad-b33f-fbee753e44cd.png)

2) what i get (`page.set_rotation(270)` , then 2-up)
![image](https://user-images.githubusercontent.com/51842098/140903567-302c82ac-1981-449f-94a9-e313fc0d3f58.png)

3) what i get (`page.set_rotation(270)`, **NO** 2-up)
2x ![image](https://user-images.githubusercontent.com/51842098/140902628-a33d2724-99d9-409b-a663-fa13b08225ed.png)

4) what i get (2-up number 3 above)
![image](https://user-images.githubusercontent.com/51842098/140903567-302c82ac-1981-449f-94a9-e313fc0d3f58.png)

**Comment by JorjMcKie** [comment]
Please send me that file - we are talking past each other.

**Reply by shin1007** [reply]
here is my pdf.
thank you for your kind support.

https://drive.google.com/open?id=1OwXxOZe_zdSNeMUSs2kipRXKqYD4BkXv&authuser=shin1007%40gmail.com&usp=drive_fs

**Comment by JorjMcKie** [comment, marked as answer]
Here is the script that does work for both those files.
Explanation:
The "rotate" argument in `show_pdf_page` can only work in a foreseeable way, if the source page is not rotated. If it is rotated, use my code snippet from a previous post.
```python
import fitz

doc1 = fitz.open("syousa2.pdf")
doc2 = fitz.open("syousa_2up.pdf")

newpdf = fitz.open()
w, h = fitz.paper_size("a3-l")

# left and right A4 rectangles:
rl = fitz.paper_rect("a4")
rr = rl + (rl.width, 0, rl.width, 0)

for page in doc1:  # loop through source pages
    if not page.number % 2:  # on even pages create the output A3 page
        rect = rl
        newpage = newpdf.new_page(width=w, height=h)
    else:
        rect = rr

    if page.rotation == 0 and page.rect.width > page.rect.height:
        rot = 90
    else:
        rot = 0

    newpage.show_pdf_page(rect, doc1, page.number, rotate=rot)

newpdf.ez_save(doc1.name + "-up.pdf")
newpdf.close()

newpdf = fitz.open()
for page in doc2:
    if not page.number % 2:
        rect = rl
        newpage = newpdf.new_page(width=w, height=h)
    else:
        rect = rr

    if page.rotation == 0 and page.rect.width > page.rect.height:
        rot = 90
    else:
        rot = 0

    newpage.show_pdf_page(rect, doc2, page.number, rotate=rot)

newpdf.ez_save(doc2.name + "-up.pdf")
```

**Reply by shin1007** [reply]
Thank you soooo much, my hero.
I'll use your snippet!!!!
